### PR TITLE
improve implementation of splitAt, take and drop.

### PR DESCRIPTION
**Pull request by JacquesCarette**
The old implementation of splitAt

```agda
splitAt : ∀ m {n} (xs : Vec A (m + n)) →
          ∃₂ λ (ys : Vec A m) (zs : Vec A n) → xs ≡ ys ++ zs
splitAt zero    xs                = ([] , xs , refl)
splitAt (suc m) (x ∷ xs)          with splitAt m xs
splitAt (suc m) (x ∷ .(ys ++ zs)) | (ys , zs , refl) =
  ((x ∷ ys) , zs , refl)
```
was strict in the proof term in the recursive call, which means that proving anything about `splitAt` was very hard.
The new one doesn't care about which proof is passed, it just uses it:
```agda
splitAt : ∀ m {n} (xs : Vec A (m + n)) →
          ∃₂ λ (ys : Vec A m) (zs : Vec A n) → xs ≡ ys ++ zs
splitAt zero    xs                = ([] , xs , refl)
splitAt (suc m) (x ∷ xs)          with splitAt m xs
splitAt (suc m) (x ∷ xs) | (ys , zs , p) = ((x ∷ ys) , zs , cong (x ∷_) p)
```

This has nice knock on effect on `take` (and `drop` as well), where the old one
```agda
take : ∀ m {n} → Vec A (m + n) → Vec A m
take m xs          with splitAt m xs
take m .(ys ++ zs) | (ys , zs , refl) = ys
```
also insisted that the proof needed to be reducible to `refl` (clearly not necessary) before just returning the first component. The new one doesn't bother to `with` at all and just returns the first component directly.

This simplified various proofs downstream which used to have to do some rewriting, but no longer need to do so.

Note that there's still green slime in this API, but that can be done later.

### CODE REVIEW

**Review by gallais — APPROVED** [review]
These changes I do agree on.

**Comment by gallais** [comment]
We will need a CHANGELOG entry.

File: src/Data/Vec/Base.agda [hunk 1]
```diff
@@ -263,16 +263,13 @@ splitAt : ∀ m {n} (xs : Vec A (m + n)) →
           ∃₂ λ (ys : Vec A m) (zs : Vec A n) → xs ≡ ys ++ zs
 splitAt zero    xs                = ([] , xs , refl)
 splitAt (suc m) (x ∷ xs)          with splitAt m xs
-splitAt (suc m) (x ∷ .(ys ++ zs)) | (ys , zs , refl) =
-  ((x ∷ ys) , zs , refl)
+splitAt (suc m) (x ∷ xs) | (ys , zs , p) = ((x ∷ ys) , zs , cong (x ∷_) p)
```
Comment (jamesmckinna) [inline src/Data/Vec/Base.agda:266]: Suggest `eq` for the proof of the equality, rather than `p`?

**Comment by jamesmckinna** [comment]
Suggest also modifying the proofs of `unfold-take` and unfold-drop` as well, in the same spirit...

File: src/Data/Vec/Base.agda (hunk 1 above)
Comment (jamesmckinna) [reply src/Data/Vec/Base.agda:266]: Suggest also using an irrefutable `with` on LHS, or even a pattern-matching `let` on RHS...

Comment (JacquesCarette) [reply src/Data/Vec/Base.agda]: both done.

**Comment by JacquesCarette** [comment]
Proofs simplified too. CHANGELOG entry forthcoming.

File: src/Data/Vec/Properties.agda [hunk 2]
```diff
@@ -215,12 +172,8 @@ lookup⇒[]= (suc i) (_ ∷ xs) p    = there (lookup⇒[]= i xs p)
...
 lookup-inject≤-take : ∀ m (m≤m+n : m ≤ m + n) (i : Fin m) (xs : Vec A (m + n)) →
                       lookup xs (Fin.inject≤ i m≤m+n) ≡ lookup (take m xs) i
-lookup-inject≤-take (suc m) m≤m+n zero (x ∷ xs)
-  rewrite unfold-take m x xs = refl
-lookup-inject≤-take (suc (suc m)) (s≤s m≤m+n) (suc zero) (x ∷ y ∷ xs)
-  rewrite unfold-take (suc m) x (y ∷ xs) | unfold-take m y xs = refl
-lookup-inject≤-take (suc (suc m)) (s≤s (s≤s m≤m+n)) (suc (suc i)) (x ∷ y ∷ xs)
-  rewrite unfold-take (suc m) x (y ∷ xs) | unfold-take m y xs = lookup-inject≤-take m m≤m+n i xs
+lookup-inject≤-take (suc m) m≤m+n zero (x ∷ xs) = refl
+lookup-inject≤-take (suc m) (s≤s m≤m+n) (suc i) (x ∷ xs) = lookup-inject≤-take m m≤m+n i xs
```
Comment (jamesmckinna) [inline src/Data/Vec/Properties.agda:176]: In the spirit of many prior PRs (and open ones eg #2000) aiming to make computation with (irrelevant) proofs of `_≤_` as lazy as possible, I'd be tempted to write this second clause as 
```agda
lookup-inject≤-take (suc m) le (suc i) (x ∷ xs) = lookup-inject≤-take m (≤-pred le) i xs
```

File: src/Data/Vec/Base.agda [hunk 3]
```diff
@@ -262,25 +262,21 @@ allFin _ = tabulate id
...
 group : ∀ n k (xs : Vec A (n * k)) →
         ∃ λ (xss : Vec (Vec A k) n) → xs ≡ concat xss
 group zero    k []                  = ([] , refl)
-group (suc n) k xs                  with splitAt k xs
-group (suc n) k .(ys ++ zs)         | (ys , zs , refl) with group n k zs
-group (suc n) k .(ys ++ concat zss) | (ys , ._ , refl) | (zss , refl) =
-  ((ys ∷ zss) , refl)
+group (suc n) k xs with splitAt k xs
+group (suc n) k xs | (ys , zs , eq) with group n k zs
+group (suc n) k xs | (ys , ._ , eq) | (zss , eq′) = ((ys ∷ zss) , trans eq (cong (ys ++_) eq′))
```
Comment (jamesmckinna) [inline src/Data/Vec/Base.agda:279]: Again, I'd consider using one of 
* nested, irrefutable, `with` here
```agda
group (suc n) k xs
  with ys , zs , eq-split ← splitAt k xs
  with zss , eq-group ← group n k zs
  =  (ys ∷ zss) , trans eq-split (cong (ys ++_) eq-group)
```
* else a pattern-matching `let`
```agda
group (suc n) k xs  = 
  let ys , zs , eq-split = splitAt k xs in
  let zss , eq-group     = group n k zs in
   (ys ∷ zss) , trans eq-split (cong (ys ++_) eq-group)
```

In each case avoiding redundant parentheses around tuples...

File: src/Data/Vec/Base.agda [hunk 4]
```diff
@@ -342,15 +338,13 @@ xs ʳ++ ys = foldl (Vec _ ∘ (_+ _)) (λ rev x → x ∷ rev) ys xs
 initLast : ∀ (xs : Vec A (1 + n)) → ∃₂ λ ys y → xs ≡ ys ∷ʳ y
 initLast {n = zero}  (x ∷ []) = ([] , x , refl)
 initLast {n = suc n} (x ∷ xs) with initLast xs
-... | (ys , y , refl) = (x ∷ ys , y , refl)
+... | (ys , y , eq) = (x ∷ ys , y , cong (x ∷_) eq)
```
Comment (jamesmckinna) [inline src/Data/Vec/Base.agda:341]: Ditto. here as to use of irrefutable `with` or `let`...

File: src/Data/Vec/Properties.agda [hunk 5]
```diff
@@ -71,86 +71,43 @@ private
...
-  ≡⟨ unfold-take m (f x y) (zipWith f xs ys) ⟩
-    f x y ∷ take m (zipWith f xs ys)
-  ≡⟨ cong (f x y ∷_) (take-distr-zipWith f xs ys) ⟩
-    f x y ∷ (zipWith f (take m xs) (take m ys))
-  ≡⟨⟩
-    zipWith f (x ∷ (take m xs)) (y ∷ (take m ys))
-  ≡˘⟨ cong₂ (zipWith f) (unfold-take m x xs) (unfold-take m y ys) ⟩
-    zipWith f (take (suc m) (x ∷ xs)) (take (suc m) (y ∷ ys))
-  ∎
+take-distr-zipWith {m = suc m} f (x ∷ xs) (y ∷ ys) = cong (f x y ∷_) (take-distr-zipWith f xs ys)
```
Comment (jamesmckinna) [inline src/Data/Vec/Properties.agda:80]: Per the style convention, should we (now, with the benefit of hindsight) rename this lemma `take-zipWith`?

File: src/Data/Vec/Properties.agda [hunk 6]
```diff
@@ -71,86 +71,43 @@ private
...
                  take m (map f xs) ≡ map f (take m xs)
 take-distr-map f zero xs = refl
-take-distr-map f (suc m) (x ∷ xs) = begin
-  take (suc m) (map f (x ∷ xs)) ≡⟨⟩
-  take (suc m) (f x ∷ map f xs) ≡⟨ unfold-take m (f x) (map f xs) ⟩
-  f x ∷ (take m (map f xs))     ≡⟨ cong (f x ∷_) (take-distr-map f m xs) ⟩
-  f x ∷ (map f (take m xs))     ≡⟨⟩
-  map f (x ∷ take m xs)         ≡˘⟨ cong (map f) (unfold-take m x xs) ⟩
-  map f (take (suc m) (x ∷ xs)) ∎
+take-distr-map f (suc m) (x ∷ xs) = cong (f x ∷_) (take-distr-map f m xs)
```
Comment (jamesmckinna) [inline src/Data/Vec/Properties.agda:85]: Ditto. `take-map`

File: src/Data/Vec/Properties.agda [hunk 7]
```diff
@@ -71,86 +71,43 @@ private
...
-  ≡⟨⟩
-    drop (suc m) (f x y ∷ (zipWith f xs ys))
-  ≡⟨ unfold-drop m (f x y) (zipWith f xs ys) ⟩
-    drop m (zipWith f xs ys)
-  ≡⟨ drop-distr-zipWith f xs ys ⟩
-    zipWith f (drop m xs) (drop m ys)
-  ≡˘⟨ cong₂ (zipWith f) (unfold-drop m x xs) (unfold-drop m y ys) ⟩
-    zipWith f (drop (suc m) (x ∷ xs)) (drop (suc m) (y ∷ ys))
-  ∎
+drop-distr-zipWith {m = suc m} f (x ∷ xs) (y ∷ ys) = drop-distr-zipWith f xs ys
```
Comment (jamesmckinna) [inline src/Data/Vec/Properties.agda:98]: `drop-zipWith`

File: src/Data/Vec/Properties.agda [hunk 8]
```diff
@@ -71,86 +71,43 @@ private
...
 drop-distr-map : ∀ (f : A → B) (m : ℕ) (x : Vec A (m + n)) →
                  drop m (map f x) ≡ map f (drop m x)
 drop-distr-map f zero x = refl
-drop-distr-map f (suc m) (x ∷ xs) = begin
-  drop (suc m) (map f (x ∷ xs)) ≡⟨⟩
-  drop (suc m) (f x ∷ map f xs) ≡⟨  unfold-drop m (f x) (map f xs) ⟩
-  drop m (map f xs)             ≡⟨  drop-distr-map f m xs ⟩
-  map f (drop m xs)             ≡˘⟨ cong (map f) (unfold-drop m x xs) ⟩
-  map f (drop (suc m) (x ∷ xs)) ∎
+drop-distr-map f (suc m) (x ∷ xs) = drop-distr-map f m xs
```
Comment (jamesmckinna) [inline src/Data/Vec/Properties.agda:103]: `drop-map`

File: src/Data/Vec/Relation/Unary/AllPairs/Properties.agda [hunk 9]
```diff
@@ -70,13 +70,11 @@ module _ {R : Rel A ℓ} where
...
   take⁺ zero pxs = []
-  take⁺ (suc m) {x ∷ xs} (px ∷ pxs)
-    rewrite Vecₚ.unfold-take m x xs = Allₚ.take⁺ m px ∷ take⁺ m pxs
+  take⁺ (suc m) {x ∷ xs} (px ∷ pxs) = Allₚ.take⁺ m px ∷ take⁺ m pxs
 
   drop⁺ : ∀ {n} m {xs} → AllPairs R {m + n} xs → AllPairs R {n} (drop m xs)
   drop⁺ zero pxs = pxs
-  drop⁺ (suc m) {x ∷ xs} (_ ∷ pxs)
-    rewrite Vecₚ.unfold-drop m x xs = drop⁺ m pxs
+  drop⁺ (suc m) {x ∷ xs} (_ ∷ pxs) = drop⁺ m pxs
```
Comment (jamesmckinna) [inline src/Data/Vec/Relation/Unary/AllPairs/Properties.agda:77]: Now that your defns permit it, I'd suppress the `{x ∷ xs}` argument

**Comment by jamesmckinna** [comment]
This all looks much nicer than before! But see some additional aesthetic tweaks to the code, for your consideration...

File: src/Data/Vec/Properties.agda (hunk 2 above)
Comment (jamesmckinna) [reply src/Data/Vec/Properties.agda:176]: Oops: computer says no! At least not until the operations on `Fin` become lazier again... (more sackcloth and ashes for me, I'm afraid)

Comment (jamesmckinna) [reply src/Data/Vec/Properties.agda:176]: Will 'resolve' this one for now, and return to it in a subsequent PR...

File: src/Data/Vec/Relation/Unary/AllPairs/Properties.agda (hunk 9 above)
Comment (JacquesCarette) [reply src/Data/Vec/Relation/Unary/AllPairs/Properties.agda:77]: Done.

File: src/Data/Vec/Base.agda (hunk 3 above)
Comment (JacquesCarette) [reply src/Data/Vec/Base.agda:279]: I like that `let` version - I'll use that one.

File: src/Data/Vec/Properties.agda (hunk 5 above)
Comment (Taneb) [reply src/Data/Vec/Properties.agda:80]: I think a) we should rename these lemmas, but b) that can be in a different PR

File: src/Data/Vec/Base.agda [hunk 10]
```diff
@@ -262,25 +262,22 @@ allFin _ = tabulate id
 splitAt : ∀ m {n} (xs : Vec A (m + n)) →
           ∃₂ λ (ys : Vec A m) (zs : Vec A n) → xs ≡ ys ++ zs
 splitAt zero    xs                = ([] , xs , refl)
-splitAt (suc m) (x ∷ xs)          with splitAt m xs
-splitAt (suc m) (x ∷ .(ys ++ zs)) | (ys , zs , refl) =
-  ((x ∷ ys) , zs , refl)
+splitAt (suc m) (x ∷ xs) =
+  let (ys , zs , eq) = splitAt m xs in ((x ∷ ys) , zs , cong (x ∷_) eq)
```
Comment (MatthewDaggitt) [inline src/Data/Vec/Base.agda:266]: No brackets necessary around `x ∷ ys`?

File: src/Data/Vec/Properties.agda (hunk 5 above)
Comment (MatthewDaggitt) [reply src/Data/Vec/Properties.agda:80]: I'm happy with either this PR or another.

Comment (gallais) [reply src/Data/Vec/Properties.agda:80]: I think every PR brings its cognitive burden. I'm not sure it's worth multiplying them for things that are not unrelated to each other.

**Comment by jamesmckinna** [comment]
Looking good to me (and thanks to @omelkonian and others at AIMXXXVI for nudging me to think harder about trade-offs between irrefutable `with` and simple(r) `let` binding...).

As for the notes about naming (I tend towards @gallais 's wish to knock these things off as you go), can you document which get changed, and which don't, if any subsequent PR should be required? I know it'll be readable in the `CHANGELOG`, but making a note here would be helpful redundancy, I think.

**Comment by jamesmckinna** [comment]
I took the liberty of resolving the merge conflicts, and removing redundant parentheses while I was at it.

File: src/Data/Vec/Properties.agda [hunk 11]
```diff
@@ -215,12 +172,8 @@ lookup⇒[]= (suc i) (_ ∷ xs) p    = there (lookup⇒[]= i xs p)
 
 lookup-inject≤-take : ∀ m (m≤m+n : m ≤ m + n) (i : Fin m) (xs : Vec A (m + n)) →
```
Comment (jamesmckinna) [inline src/Data/Vec/Properties.agda:173]: Explicit argument `m` could be taken to be implicit here? It's inferrable from the form of `i`...

File: src/Data/Vec/Properties.agda (hunk 5 above)
Comment (jamesmckinna) [reply src/Data/Vec/Properties.agda]: Done, modulo `CHANGELOG`.

**Comment by jamesmckinna** [comment]
Re: `CHANGELOG` @gallais what is needed besides the deprecations of the `*-distr-*` names, given that the major contribution is merely to streamline a definition... no types are changed by this, for example. So where/how do we account for this. Is it even a breaking/non-backwards-compatible change?

**Comment by gallais** [comment]
Yes you're right, the only entry we need is the one about the renamings.

**Comment by jamesmckinna** [comment]
Done.

**Comment by jamesmckinna** [comment]
Last thoughts: why do we not also consider, in `Data.List.Base` the following improved versions of `initLast`/`unsnoc`:
```agda
initLast : ∀ (xs : List A) → Maybe (∃₂ λ ys y → xs ≡ ys ∷ʳ y)
initLast []               = nothing
initLast (x ∷ [])         = just ([] , x , refl)
initLast (x ∷ xs@(_ ∷ _)) =
  maybe′ (λ (ys , y , eq) → just (x ∷ ys , y , cong (x ∷_) eq)) nothing (initLast xs)

unsnoc : List A → Maybe (List A × A)
unsnoc as = maybe′ (λ (ys , y , _) → just (x ∷ ys , y) nothing (initLast as)
```
with an optimisation if we were also to consider `NonNull` as an irrelevant instance predicate for `List`s with the obvious definition analogous to `NonZero`...
```agda
initLast : ∀ (xs : List A) .{{_ : NonNull xs}} → ∃₂ λ ys y → xs ≡ ys ∷ʳ y
initLast (x ∷ [])         = just ([] , x , refl)
initLast (x ∷ xs@(_ ∷ _)) = let ys , y , eq = initLast xs in x ∷ ys , y , cong (x ∷_) eq
```
... just sayin'... for another time ;-)

**Pull request merged**